### ISSUE TITLE
WAMP-cryptosign: signature primitive extensibility

### ISSUE THREAD

**Issue by oberstet** [wamp-proto/wamp-proto#298]
WAMP-cryptosign is an asymmetric, signature based authentication scheme that is defined by two parts:

1. the message exchange pattern (authentication handshake) and construction of challenge to sign
2. the choice of signature algorithm (the cryptographic primitive at the core)

Currently, the algorithm used for signing challenges sent by the router to authenticate a connecting client is fixed: Ed25519.

Ed25519 has a number of attractive features, like elliptic curve based (efficient), short keys and signature, high-quality reference implementation (NaCl), and so on.

However, Ed25519 being based on elliptic curves is expected to be attackable eg using quantum computers, but there are signature schemes that are [expected to be resistent](https://en.wikipedia.org/wiki/Post-quantum_cryptography).

[liboqs](https://github.com/open-quantum-safe/liboqs/) is a C library for quantum-resistant cryptographic algorithms. It exposes a [signature API](https://github.com/open-quantum-safe/liboqs/blob/master/src/sig/sig.h) with a parameter to select one of multiple signature algorithms:

* OQS_SIG_picnic_L1_FS,
* OQS_SIG_picnic_L1_UR,
* OQS_SIG_picnic_L3_FS,
* OQS_SIG_picnic_L3_UR,
* OQS_SIG_picnic_L5_FS,
* OQS_SIG_picnic_L5_UR,

---

The proposal here is to expand WAMP-cryptosign, eg:

* `wamp-cryptosign`
* `wamp-cryptosign-l1fs`
* `wamp-cryptosign-l1ur`
* `wamp-cryptosign-l3fs`
* ...

**Comment by oberstet** [comment]
another scheme to look at: [XMSS](https://datatracker.ietf.org/doc/draft-irtf-cfrg-xmss-hash-based-signatures/). This however, is far less in code development, eg the [reference implementation](https://github.com/joostrijneveld/xmss-reference).

**Comment by konsultaner** [comment]
Why not add an `algorithm` to the `authextra` next to the `pubkey` field?

``` 
{
    'channel_binding': 'null',
    'pubkey': 'hajfhjashfkjsa==',
    'algorithm': 'eddsa'  // or 'l1fs'
}
```

**Comment by oberstet** [comment]
actually, yes, that should do the trick as well! and it is backwards compatible (we can define non-existent `algorithm` field as `eddsa` ..).

so what values could we use? here is a proposal:

1. **algorithm=="ed25519"** (*default*): "Ed25519 is the EdDSA signature scheme using SHA-512 (SHA-2) and Curve25519"
2. **algorithm=="ed448"** (*new*): "Ed448 is the EdDSA signature scheme using SHAKE256 (SHA-3) and Curve448 defined in RFC 8032. It has also been approved in the draft of the FIPS 186-5 standard."

others?
maybe stuff from https://github.com/open-quantum-safe/liboqs/#signature-schemes ?

**Comment by konsultaner** [comment]
Maybe  this is also a good source to look at:

https://csrc.nist.gov/Projects/post-quantum-cryptography/round-3-submissions

**Comment by oberstet** [comment]
^ yes, this makes sense!

for WAMP-cryptosign, we need category "Digital Signature Algorithms" ..  that is:

Round 3 Finalists: Digital Signature Algorithms:
* CRYSTALS-DILITHIUM 	
* FALCON 	
* Rainbow

https://pq-crystals.org/dilithium/index.shtml
https://csrc.nist.gov/CSRC/media/Presentations/crystals-dilithium-round-2-update/images-media/crystals-dilithium-lyubashevsky.pdf

**Comment by konsultaner** [comment]
@oberstet Just one other thought. As far as I know, there is no spec for e2e encryption yet, right? Wouldn't it be nice to use the `pubkey` ,etc. to support e2e encryption as well? Maybe by adding another flag `e2e: true` or what ever makes sense?

**Comment by oberstet** [comment]
yes, there is no spec about the end-to-end encryption stuff yet. 

historically, the feature roots back to an issue raised by a user:

https://github.com/wamp-proto/wamp-proto/issues/81

this triggered the development of a prototype in Autobahn that does application payload e2e encryption

https://github.com/crossbario/crossbar-examples/tree/master/encryption/cryptobox

during the implementation it became clear that another feature "payload transparency" would be good (because with e2e, the router cannot read the app payload anyways)

https://github.com/wamp-proto/wamp-proto/issues/228

then, later on, we recognized that the e2e key management, when expanded with "selling/buying keys" via an intermediary (the market maker), and by anchoring everything trust-wise in the blockchain, including the buy/sell data service (WAMP API usage) transactions (the market maker act as an off-chain key exchange with the transactions happening off-chain in so-called state channels, but rooted on-chain in smart contracts), could serve a much further reaching vision:

https://xbr.network/

from an app developer perspective, this then looks like this

https://github.com/crossbario/crossbar-examples/tree/master/xbr/teststack1/python

----

all of above are closely related, but "alpha", and not written down in specs ... but I would be highly interested in discussing this in the community, possibly modifying it, simplying and throwing away parts that are not needed / redundant.

I am quite heavily invested in this stuff .. have been working on it essentially the last 2 years. I also think that this in a way brings the original vision for WAMP to a natural conclusion.

strong authentication, and strong end-to-end user data encryption. with blockchain trust anchors and off-chain data transactions.

---

rgd using _one_ key for

* session authentication and
* end-to-end data encryption and
* transaction signing ("buy/sell" key, even with price 0).

has 2 problems:

* it is generally discouraged to have keys with multiple uses (authentication, encryption, signing, ..)
* it is a practical problem as the key algos differ between WAMP-cryptosign and Ethereum, and my attempts to add it failed: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-665.md

**Comment by konsultaner** [comment]
@oberstet Thank you for that very detailed description. I will have to dig a little deeper into that topic to be able to discuss it. I'll start reading into that topic soon, but I want to finish the ed25519 key loading from different sources for cryptosign first.
